<a href="https://colab.research.google.com/github/Sanjiwani1wadhave/Assignment-12/blob/main/Navive_Bayes_Assignment_12.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Import Libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [ ]:
import warnings

warnings.filterwarnings('ignore')

In [ ]:
# Import datasets
from google.colab import files
uploaded= files.upload()

Saving SalaryData_Test.csv.crdownload to SalaryData_Test.csv.crdownload


In [ ]:
salary_test=pd.read_csv("SalaryData_Test.csv.crdownload")
salary_test.head()

age   workclass      education  educationno        maritalstatus  \
0   25     Private           11th            7        Never-married   
1   38     Private        HS-grad            9   Married-civ-spouse   
2   28   Local-gov     Assoc-acdm           12   Married-civ-spouse   
3   44     Private   Some-college           10   Married-civ-spouse   
4   34     Private           10th            6        Never-married   

           occupation    relationship    race    sex  capitalgain  \
0   Machine-op-inspct       Own-child   Black   Male          0.0   
1     Farming-fishing         Husband   White   Male          0.0   
2     Protective-serv         Husband   White   Male          0.0   
3   Machine-op-inspct         Husband   Black   Male       7688.0   
4       Other-service   Not-in-family   White   Male          0.0   

   capitalloss  hoursperweek          native  Salary  
0          0.0          40.0   United-States   <=50K  
1          0.0          50.0   United-States   <=50K  
2          0.0          40.0   United-States    >50K  
3          0.0          40.0   United-States    >50K  
4          0.0          30.0   United-States   <=50K

In [ ]:
from google.colab import files
uploaded=files.upload()

Saving SalaryData_Train.csv to SalaryData_Train.csv


In [ ]:
salary_train=pd.read_csv("SalaryData_Train.csv")
salary_train

age          workclass    education  educationno        maritalstatus  \
0       39          State-gov    Bachelors           13        Never-married   
1       50   Self-emp-not-inc    Bachelors           13   Married-civ-spouse   
2       38            Private      HS-grad            9             Divorced   
3       53            Private         11th            7   Married-civ-spouse   
4       28            Private    Bachelors           13   Married-civ-spouse   
...    ...                ...          ...          ...                  ...   
30156   27            Private   Assoc-acdm           12   Married-civ-spouse   
30157   40            Private      HS-grad            9   Married-civ-spouse   
30158   58            Private      HS-grad            9              Widowed   
30159   22            Private      HS-grad            9        Never-married   
30160   52       Self-emp-inc      HS-grad            9   Married-civ-spouse   

               occupation    relationship    race      sex  capitalgain  \
0            Adm-clerical   Not-in-family   White     Male         2174   
1         Exec-managerial         Husband   White     Male            0   
2       Handlers-cleaners   Not-in-family   White     Male            0   
3       Handlers-cleaners         Husband   Black     Male            0   
4          Prof-specialty            Wife   Black   Female            0   
...                   ...             ...     ...      ...          ...   
30156        Tech-support            Wife   White   Female            0   
30157   Machine-op-inspct         Husband   White     Male            0   
30158        Adm-clerical       Unmarried   White   Female            0   
30159        Adm-clerical       Own-child   White     Male            0   
30160     Exec-managerial            Wife   White   Female        15024   

       capitalloss  hoursperweek          native  Salary  
0                0            40   United-States   <=50K  
1                0            13   United-States   <=50K  
2                0            40   United-States   <=50K  
3                0            40   United-States   <=50K  
4                0            40            Cuba   <=50K  
...            ...           ...             ...     ...  
30156            0            38   United-States   <=50K  
30157            0            40   United-States    >50K  
30158            0            40   United-States   <=50K  
30159            0            20   United-States   <=50K  
30160            0            40   United-States    >50K  

[30161 rows x 14 columns]

**EDA**

In [ ]:
salary_test.shape

(6548, 14)

In [ ]:
salary_train.shape

(30161, 14)

In [ ]:
#Rename column names
col_names=['age', 'workclass', 'education', 'education_num', 'marital_status', 'occupation', 'relationship', 'race', 'sex', 'capital_gain', 'capital_loss', 'hours_per_week', 'native_country', 'salary']
salary_test.columns=col_names
salary_train.columns=col_names


In [ ]:
salary_test.columns

Index(['age', 'workclass', 'education', 'education_num', 'marital_status',
       'occupation', 'relationship', 'race', 'sex', 'capital_gain',
       'capital_loss', 'hours_per_week', 'native_country', 'salary'],
      dtype='object')

In [ ]:
salary_train.columns

Index(['age', 'workclass', 'education', 'education_num', 'marital_status',
       'occupation', 'relationship', 'race', 'sex', 'capital_gain',
       'capital_loss', 'hours_per_week', 'native_country', 'salary'],
      dtype='object')

In [ ]:
salary_test.head()

age   workclass      education  education_num       marital_status  \
0   25     Private           11th              7        Never-married   
1   38     Private        HS-grad              9   Married-civ-spouse   
2   28   Local-gov     Assoc-acdm             12   Married-civ-spouse   
3   44     Private   Some-college             10   Married-civ-spouse   
4   34     Private           10th              6        Never-married   

           occupation    relationship    race    sex  capital_gain  \
0   Machine-op-inspct       Own-child   Black   Male           0.0   
1     Farming-fishing         Husband   White   Male           0.0   
2     Protective-serv         Husband   White   Male           0.0   
3   Machine-op-inspct         Husband   Black   Male        7688.0   
4       Other-service   Not-in-family   White   Male           0.0   

   capital_loss  hours_per_week  native_country  salary  
0           0.0            40.0   United-States   <=50K  
1           0.0            50.0   United-States   <=50K  
2           0.0            40.0   United-States    >50K  
3           0.0            40.0   United-States    >50K  
4           0.0            30.0   United-States   <=50K

In [ ]:
salary_train.head()

age          workclass   education  education_num       marital_status  \
0   39          State-gov   Bachelors             13        Never-married   
1   50   Self-emp-not-inc   Bachelors             13   Married-civ-spouse   
2   38            Private     HS-grad              9             Divorced   
3   53            Private        11th              7   Married-civ-spouse   
4   28            Private   Bachelors             13   Married-civ-spouse   

           occupation    relationship    race      sex  capital_gain  \
0        Adm-clerical   Not-in-family   White     Male          2174   
1     Exec-managerial         Husband   White     Male             0   
2   Handlers-cleaners   Not-in-family   White     Male             0   
3   Handlers-cleaners         Husband   Black     Male             0   
4      Prof-specialty            Wife   Black   Female             0   

   capital_loss  hours_per_week  native_country  salary  
0             0              40   United-States   <=50K  
1             0              13   United-States   <=50K  
2             0              40   United-States   <=50K  
3             0              40   United-States   <=50K  
4             0              40            Cuba   <=50K

**Summary of data**

In [ ]:
salary_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6548 entries, 0 to 6547
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             6548 non-null   int64  
 1   workclass       6548 non-null   object 
 2   education       6548 non-null   object 
 3   education_num   6548 non-null   int64  
 4   marital_status  6548 non-null   object 
 5   occupation      6547 non-null   object 
 6   relationship    6547 non-null   object 
 7   race            6547 non-null   object 
 8   sex             6547 non-null   object 
 9   capital_gain    6547 non-null   float64
 10  capital_loss    6547 non-null   float64
 11  hours_per_week  6547 non-null   float64
 12  native_country  6547 non-null   object 
 13  salary          6547 non-null   object 
dtypes: float64(3), int64(2), object(9)
memory usage: 716.3+ KB


In [ ]:
salary_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30161 entries, 0 to 30160
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             30161 non-null  int64 
 1   workclass       30161 non-null  object
 2   education       30161 non-null  object
 3   education_num   30161 non-null  int64 
 4   marital_status  30161 non-null  object
 5   occupation      30161 non-null  object
 6   relationship    30161 non-null  object
 7   race            30161 non-null  object
 8   sex             30161 non-null  object
 9   capital_gain    30161 non-null  int64 
 10  capital_loss    30161 non-null  int64 
 11  hours_per_week  30161 non-null  int64 
 12  native_country  30161 non-null  object
 13  salary          30161 non-null  object
dtypes: int64(5), object(9)
memory usage: 3.2+ MB


**Explore Categorical Variables**

In [ ]:
# Find Categorical variables
categorical = [var for var in salary_test.columns if salary_test[var].dtype=='O']

print('There are {} categorical variables\n'.format(len(categorical)))

print('The categorical variables are :\n\n', categorical)

There are 9 categorical variables

The categorical variables are :

 ['workclass', 'education', 'marital_status', 'occupation', 'relationship', 'race', 'sex', 'native_country', 'salary']


In [ ]:
categorical = [var for var in salary_train.columns if salary_train[var].dtype=='O']

print('There are {} categorical variables\n'.format(len(categorical)))

print('The categorical variables are :\n\n', categorical)

There are 9 categorical variables

The categorical variables are :

 ['workclass', 'education', 'marital_status', 'occupation', 'relationship', 'race', 'sex', 'native_country', 'salary']


In [ ]:
# view the categorical variables
salary_train[categorical].head()

workclass   education       marital_status          occupation  \
0          State-gov   Bachelors        Never-married        Adm-clerical   
1   Self-emp-not-inc   Bachelors   Married-civ-spouse     Exec-managerial   
2            Private     HS-grad             Divorced   Handlers-cleaners   
3            Private        11th   Married-civ-spouse   Handlers-cleaners   
4            Private   Bachelors   Married-civ-spouse      Prof-specialty   

     relationship    race      sex  native_country  salary  
0   Not-in-family   White     Male   United-States   <=50K  
1         Husband   White     Male   United-States   <=50K  
2   Not-in-family   White     Male   United-States   <=50K  
3         Husband   Black     Male   United-States   <=50K  
4            Wife   Black   Female            Cuba   <=50K

In [ ]:
salary_test[categorical].head()

workclass      education       marital_status          occupation  \
0     Private           11th        Never-married   Machine-op-inspct   
1     Private        HS-grad   Married-civ-spouse     Farming-fishing   
2   Local-gov     Assoc-acdm   Married-civ-spouse     Protective-serv   
3     Private   Some-college   Married-civ-spouse   Machine-op-inspct   
4     Private           10th        Never-married       Other-service   

     relationship    race    sex  native_country  salary  
0       Own-child   Black   Male   United-States   <=50K  
1         Husband   White   Male   United-States   <=50K  
2         Husband   White   Male   United-States    >50K  
3         Husband   Black   Male   United-States    >50K  
4   Not-in-family   White   Male   United-States   <=50K

**Missing values in the categorical variables**

In [ ]:
salary_test[categorical].isnull().sum()

workclass         0
education         0
marital_status    0
occupation        1
relationship      1
race              1
sex               1
native_country    1
salary            1
dtype: int64

In [ ]:
salary_train[categorical].isnull().sum()

workclass         0
education         0
marital_status    0
occupation        0
relationship      0
race              0
sex               0
native_country    0
salary            0
dtype: int64

**Frequency counts of categorical variables**

In [ ]:
for var in categorical:

  print(salary_test[var].value_counts())

 Private             4774
 Self-emp-not-inc     544
 Local-gov            447
 State-gov            316
 Self-emp-inc         264
 Federal-gov          200
 Without-pay            3
Name: workclass, dtype: int64
 HS-grad         2113
 Some-college    1420
 Bachelors       1114
 Masters          381
 Assoc-voc        287
 11th             247
 Assoc-acdm       201
 10th             189
 7th-8th          125
 Prof-school      100
 9th              100
 12th              86
 Doctorate         79
 5th-6th           65
 1st-4th           30
 Preschool         11
Name: education, dtype: int64
 Married-civ-spouse       2960
 Never-married            2165
 Divorced                  922
 Widowed                   209
 Separated                 206
 Married-spouse-absent      78
 Married-AF-spouse           7
 Married                     1
Name: marital_status, dtype: int64
 Exec-managerial      885
 Prof-specialty       872
 Craft-repair         845
 Adm-clerical         789
 Sales             

In [ ]:
for var in categorical:

  print(salary_train[var].value_counts())

 Private             22285
 Self-emp-not-inc     2499
 Local-gov            2067
 State-gov            1279
 Self-emp-inc         1074
 Federal-gov           943
 Without-pay            14
Name: workclass, dtype: int64
 HS-grad         9840
 Some-college    6677
 Bachelors       5044
 Masters         1627
 Assoc-voc       1307
 11th            1048
 Assoc-acdm      1008
 10th             820
 7th-8th          557
 Prof-school      542
 9th              455
 12th             377
 Doctorate        375
 5th-6th          288
 1st-4th          151
 Preschool         45
Name: education, dtype: int64
 Married-civ-spouse       14065
 Never-married             9725
 Divorced                  4214
 Separated                  939
 Widowed                    827
 Married-spouse-absent      370
 Married-AF-spouse           21
Name: marital_status, dtype: int64
 Prof-specialty       4038
 Craft-repair         4030
 Exec-managerial      3992
 Adm-clerical         3721
 Sales                3584
 Othe

In [ ]:
# view frequency distribution of categorical variables

for var in categorical:

    print(salary_test[var].value_counts()/np.float(len(salary_test)))


 Private             0.729078
 Self-emp-not-inc    0.083079
 Local-gov           0.068265
 State-gov           0.048259
 Self-emp-inc        0.040318
 Federal-gov         0.030544
 Without-pay         0.000458
Name: workclass, dtype: float64
 HS-grad         0.322694
 Some-college    0.216860
 Bachelors       0.170128
 Masters         0.058186
 Assoc-voc       0.043830
 11th            0.037721
 Assoc-acdm      0.030696
 10th            0.028864
 7th-8th         0.019090
 Prof-school     0.015272
 9th             0.015272
 12th            0.013134
 Doctorate       0.012065
 5th-6th         0.009927
 1st-4th         0.004582
 Preschool       0.001680
Name: education, dtype: float64
 Married-civ-spouse       0.452046
 Never-married            0.330635
 Divorced                 0.140806
 Widowed                  0.031918
 Separated                0.031460
 Married-spouse-absent    0.011912
 Married-AF-spouse        0.001069
 Married                  0.000153
Name: marital_status, dtype: f

In [ ]:
# view frequency distribution of categorical variables

for var in categorical:

    print(salary_train[var].value_counts()/np.float(len(salary_train)))


 Private             0.738868
 Self-emp-not-inc    0.082855
 Local-gov           0.068532
 State-gov           0.042406
 Self-emp-inc        0.035609
 Federal-gov         0.031266
 Without-pay         0.000464
Name: workclass, dtype: float64
 HS-grad         0.326249
 Some-college    0.221379
 Bachelors       0.167236
 Masters         0.053944
 Assoc-voc       0.043334
 11th            0.034747
 Assoc-acdm      0.033421
 10th            0.027187
 7th-8th         0.018468
 Prof-school     0.017970
 9th             0.015086
 12th            0.012500
 Doctorate       0.012433
 5th-6th         0.009549
 1st-4th         0.005006
 Preschool       0.001492
Name: education, dtype: float64
 Married-civ-spouse       0.466331
 Never-married            0.322436
 Divorced                 0.139717
 Separated                0.031133
 Widowed                  0.027420
 Married-spouse-absent    0.012267
 Married-AF-spouse        0.000696
Name: marital_status, dtype: float64
 Prof-specialty       0.1338

**Explore workclass variable**

In [ ]:
# check labels in workclass variables

salary_test.workclass.unique()

array([' Private', ' Local-gov', ' Self-emp-not-inc', ' Federal-gov',
       ' State-gov', ' Self-emp-inc', ' Without-pay'], dtype=object)

In [ ]:
salary_train.workclass.unique()

array([' State-gov', ' Self-emp-not-inc', ' Private', ' Federal-gov',
       ' Local-gov', ' Self-emp-inc', ' Without-pay'], dtype=object)

In [ ]:
# check frequency distribution of values in workclass variable
salary_test.workclass.value_counts()

 Private             4774
 Self-emp-not-inc     544
 Local-gov            447
 State-gov            316
 Self-emp-inc         264
 Federal-gov          200
 Without-pay            3
Name: workclass, dtype: int64

In [ ]:
salary_train.workclass.value_counts()

 Private             22285
 Self-emp-not-inc     2499
 Local-gov            2067
 State-gov            1279
 Self-emp-inc         1074
 Federal-gov           943
 Without-pay            14
Name: workclass, dtype: int64

**Explore occupation variables**

In [ ]:
# check labels in occupation variable

salary_test.occupation.unique()

array([' Machine-op-inspct', ' Farming-fishing', ' Protective-serv',
       ' Other-service', ' Prof-specialty', ' Craft-repair',
       ' Adm-clerical', ' Exec-managerial', ' Tech-support', ' Sales',
       ' Priv-house-serv', ' Transport-moving', ' Handlers-cleaners',
       ' Armed-Forces', nan], dtype=object)

In [ ]:
salary_train.occupation.unique()

array([' Adm-clerical', ' Exec-managerial', ' Handlers-cleaners',
       ' Prof-specialty', ' Other-service', ' Sales', ' Transport-moving',
       ' Farming-fishing', ' Machine-op-inspct', ' Tech-support',
       ' Craft-repair', ' Protective-serv', ' Armed-Forces',
       ' Priv-house-serv'], dtype=object)

In [ ]:
# check frequency distribution of values in occupation variable

salary_test.occupation.value_counts()

 Exec-managerial      885
 Prof-specialty       872
 Craft-repair         845
 Adm-clerical         789
 Sales                781
 Other-service        697
 Machine-op-inspct    461
 Transport-moving     299
 Handlers-cleaners    299
 Farming-fishing      223
 Tech-support         201
 Protective-serv      145
 Priv-house-serv       48
 Armed-Forces           2
Name: occupation, dtype: int64

In [ ]:

salary_train.occupation.value_counts()

 Prof-specialty       4038
 Craft-repair         4030
 Exec-managerial      3992
 Adm-clerical         3721
 Sales                3584
 Other-service        3212
 Machine-op-inspct    1965
 Transport-moving     1572
 Handlers-cleaners    1350
 Farming-fishing       989
 Tech-support          912
 Protective-serv       644
 Priv-house-serv       143
 Armed-Forces            9
Name: occupation, dtype: int64

**Explore Native-Country Variable**

In [ ]:
# check labels in native_country variable

salary_test.native_country.unique()

array([' United-States', ' Peru', ' Guatemala', ' Mexico',
       ' Dominican-Republic', ' Ireland', ' Germany', ' Philippines',
       ' Thailand', ' Haiti', ' El-Salvador', ' Puerto-Rico', ' Vietnam',
       ' South', ' Columbia', ' Japan', ' India', ' Cambodia', ' Poland',
       ' Laos', ' England', ' Cuba', ' Taiwan', ' Italy', ' Canada',
       ' Portugal', ' China', ' Nicaragua', ' Honduras', ' Iran',
       ' Scotland', ' Jamaica', ' Ecuador', ' Yugoslavia', ' Hungary',
       ' Hong', ' Greece', ' Trinadad&Tobago',
       ' Outlying-US(Guam-USVI-etc)', ' France', nan], dtype=object)

In [ ]:
salary_train.native_country.unique()

array([' United-States', ' Cuba', ' Jamaica', ' India', ' Mexico',
       ' Puerto-Rico', ' Honduras', ' England', ' Canada', ' Germany',
       ' Iran', ' Philippines', ' Poland', ' Columbia', ' Cambodia',
       ' Thailand', ' Ecuador', ' Laos', ' Taiwan', ' Haiti', ' Portugal',
       ' Dominican-Republic', ' El-Salvador', ' France', ' Guatemala',
       ' Italy', ' China', ' South', ' Japan', ' Yugoslavia', ' Peru',
       ' Outlying-US(Guam-USVI-etc)', ' Scotland', ' Trinadad&Tobago',
       ' Greece', ' Nicaragua', ' Vietnam', ' Hong', ' Ireland',
       ' Hungary'], dtype=object)

In [ ]:
# check frequency distribution of values in native_country variable

salary_test.native_country.value_counts()

 United-States                 6010
 Mexico                         136
 Philippines                     41
 Puerto-Rico                     30
 Germany                         27
 India                           24
 Canada                          23
 El-Salvador                     20
 England                         18
 Cuba                            17
 Guatemala                       15
 China                           13
 Dominican-Republic              12
 South                           12
 Poland                          11
 Haiti                           11
 Columbia                        11
 Portugal                        10
 Italy                           10
 Jamaica                          9
 Japan                            9
 Thailand                         8
 Nicaragua                        7
 Vietnam                          6
 Ecuador                          6
 Taiwan                           5
 Peru                             5
 Greece                     

In [ ]:
salary_train.native_country.value_counts()

 United-States                 27504
 Mexico                          610
 Philippines                     188
 Germany                         128
 Puerto-Rico                     109
 Canada                          107
 India                           100
 El-Salvador                     100
 Cuba                             92
 England                          86
 Jamaica                          80
 South                            71
 China                            68
 Italy                            68
 Dominican-Republic               67
 Vietnam                          64
 Guatemala                        63
 Japan                            59
 Poland                           56
 Columbia                         56
 Iran                             42
 Taiwan                           42
 Haiti                            42
 Portugal                         34
 Nicaragua                        33
 Peru                             30
 Greece                           29
 

In [ ]:
salary_test[categorical].isnull().sum()

workclass         0
education         0
marital_status    0
occupation        1
relationship      1
race              1
sex               1
native_country    1
salary            1
dtype: int64

Now, we can see that occupation, relationship, race,sex, native_country, salary variable contains missing values.

In [ ]:
# check for cardinality in categorical variables

for var in categorical:

    print(var, ' contains ', len(salary_test[var].unique()), ' labels')

workclass  contains  7  labels
education  contains  16  labels
marital_status  contains  8  labels
occupation  contains  15  labels
relationship  contains  7  labels
race  contains  6  labels
sex  contains  3  labels
native_country  contains  41  labels
salary  contains  3  labels



We can see that native_country column contains relatively large number of labels as compared to other columns.

In [ ]:


for var in categorical:

    print(var, ' contains ', len(salary_train[var].unique()), ' labels')

workclass  contains  7  labels
education  contains  16  labels
marital_status  contains  7  labels
occupation  contains  14  labels
relationship  contains  6  labels
race  contains  5  labels
sex  contains  2  labels
native_country  contains  40  labels
salary  contains  2  labels



We can see that native_country column contains relatively large number of labels as compared to other columns.

**Explore Numerical Variables**

In [ ]:
# find numerical variables

numerical = [var for var in salary_test.columns if salary_test[var].dtype!='O']

print('There are {} numerical variables\n'.format(len(numerical)))

print('The numerical variables are :', numerical)

There are 5 numerical variables

The numerical variables are : ['age', 'education_num', 'capital_gain', 'capital_loss', 'hours_per_week']


In [ ]:


numerical = [var for var in salary_train.columns if salary_train[var].dtype!='O']

print('There are {} numerical variables\n'.format(len(numerical)))

print('The numerical variables are :', numerical)

There are 5 numerical variables

The numerical variables are : ['age', 'education_num', 'capital_gain', 'capital_loss', 'hours_per_week']


In [ ]:
# view the numerical variables

salary_test[numerical].head()

age  education_num  capital_gain  capital_loss  hours_per_week
0   25              7           0.0           0.0            40.0
1   38              9           0.0           0.0            50.0
2   28             12           0.0           0.0            40.0
3   44             10        7688.0           0.0            40.0
4   34              6           0.0           0.0            30.0

In [ ]:
salary_train[numerical].head()

age  education_num  capital_gain  capital_loss  hours_per_week
0   39             13          2174             0              40
1   50             13             0             0              13
2   38              9             0             0              40
3   53              7             0             0              40
4   28             13             0             0              40

**Missing values in numerical variables**

In [ ]:
# check missing values in numerical variables

salary_test[numerical].isnull().sum()

age               0
education_num     0
capital_gain      1
capital_loss      1
hours_per_week    1
dtype: int64

In [ ]:
salary_train[numerical].isnull().sum()

age               0
education_num     0
capital_gain      0
capital_loss      0
hours_per_week    0
dtype: int64

**Declare feature vector and target variable**

In [ ]:
X1 = salary_train.drop(['salary'], axis=1)

y1= salary_train['salary']

In [ ]:
X2 = salary_test.drop(['salary'], axis=1)

y2= salary_test['salary']

**Split data into separate training and testing set**

In [ ]:
# split X and y into training and testing sets

from sklearn.model_selection import train_test_split

X1_train, X1_test, y1_train, y1_test = train_test_split(X1, y1, test_size = 0.2, random_state = 0)

In [ ]:
# check the shape of X_train and X_test

X1_train.shape, X1_test.shape

((24128, 13), (6033, 13))

In [ ]:
# split X and y into training and testing sets

from sklearn.model_selection import train_test_split

X2_train, X2_test, y2_train, y2_test = train_test_split(X2, y2, test_size = 0.2, random_state = 0)

In [ ]:
# check the shape of X_train and X_test

X2_train.shape, X2_test.shape

((5238, 13), (1310, 13))

# **Feature Engineering**

In [ ]:
# check data types in X_train

X1_train.dtypes

age                int64
workclass         object
education         object
education_num      int64
marital_status    object
occupation        object
relationship      object
race              object
sex               object
capital_gain       int64
capital_loss       int64
hours_per_week     int64
native_country    object
dtype: object

In [ ]:
# display categorical variables

categorical = [col for col in X1_train.columns if X1_train[col].dtypes == 'O']

categorical

['workclass',
 'education',
 'marital_status',
 'occupation',
 'relationship',
 'race',
 'sex',
 'native_country']

In [ ]:
# check data types in X_train

X2_train.dtypes

age                 int64
workclass          object
education          object
education_num       int64
marital_status     object
occupation         object
relationship       object
race               object
sex                object
capital_gain      float64
capital_loss      float64
hours_per_week    float64
native_country     object
dtype: object

In [ ]:
# display categorical variables

categorical = [col for col in X2_train.columns if X2_train[col].dtypes == 'O']

categorical

['workclass',
 'education',
 'marital_status',
 'occupation',
 'relationship',
 'race',
 'sex',
 'native_country']

In [ ]:
# display numerical variables

numerical = [col for col in X1_train.columns if X1_train[col].dtypes != 'O']

numerical

['age', 'education_num', 'capital_gain', 'capital_loss', 'hours_per_week']

In [ ]:



numerical = [col for col in X2_train.columns if X2_train[col].dtypes != 'O']

numerical

['age', 'education_num', 'capital_gain', 'capital_loss', 'hours_per_week']

**Engineering missing values in categorical variables**

In [ ]:
X1_train[categorical].isnull().mean()

workclass         0.0
education         0.0
marital_status    0.0
occupation        0.0
relationship      0.0
race              0.0
sex               0.0
native_country    0.0
dtype: float64

In [ ]:
X2_train[categorical].isnull().mean()

workclass         0.000000
education         0.000000
marital_status    0.000000
occupation        0.000191
relationship      0.000191
race              0.000191
sex               0.000191
native_country    0.000191
dtype: float64

In [ ]:
# print categorical variables with missing data

for col in categorical:
    if X2_train[col].isnull().mean()>0:
        print(col, (X2_train[col].isnull().mean()))

occupation 0.00019091256204658267
relationship 0.00019091256204658267
race 0.00019091256204658267
sex 0.00019091256204658267
native_country 0.00019091256204658267


In [ ]:
# impute missing categorical variables with most frequent value

for salary_train2 in [X2_train, X2_test]:
    salary_train2['race'].fillna(X2_train['race'].mode()[0], inplace=True)
    salary_train2['occupation'].fillna(X2_train['occupation'].mode()[0], inplace=True)
    salary_train2['native_country'].fillna(X2_train['native_country'].mode()[0], inplace=True)
    salary_train2['sex'].fillna(X2_train['sex'].mode()[0], inplace=True)
    salary_train2['relationship'].fillna(X2_train['relationship'].mode()[0], inplace=True)

In [ ]:
# check missing values in categorical variables in X_train

X1_train[categorical].isnull().sum()

workclass         0
education         0
marital_status    0
occupation        0
relationship      0
race              0
sex               0
native_country    0
dtype: int64

In [ ]:
X2_train[categorical].isnull().sum()

workclass         0
education         0
marital_status    0
occupation        0
relationship      0
race              0
sex               0
native_country    0
dtype: int64

As a final check, I will check for missing values in X_train and X_test.

In [ ]:
# check missing values in categorical variables in X_test

X1_test[categorical].isnull().sum()

workclass         0
education         0
marital_status    0
occupation        0
relationship      0
race              0
sex               0
native_country    0
dtype: int64

In [ ]:
# check missing values in X_train

X1_train.isnull().sum()

age               0
workclass         0
education         0
education_num     0
marital_status    0
occupation        0
relationship      0
race              0
sex               0
capital_gain      0
capital_loss      0
hours_per_week    0
native_country    0
dtype: int64

In [ ]:
# check missing values in X_test

X2_test.isnull().sum()

age               0
workclass         0
education         0
education_num     0
marital_status    0
occupation        0
relationship      0
race              0
sex               0
capital_gain      0
capital_loss      0
hours_per_week    0
native_country    0
dtype: int64

In [ ]:
X2_train.isnull().sum()

age               0
workclass         0
education         0
education_num     0
marital_status    0
occupation        0
relationship      0
race              0
sex               0
capital_gain      1
capital_loss      1
hours_per_week    1
native_country    0
dtype: int64

In [ ]:
# impute missing categorical variables with most frequent value

for salary_train2 in [X2_train, X2_test]:
    salary_train2['capital_gain'].fillna(X2_train['capital_gain'].mode()[0], inplace=True)
    salary_train2['capital_loss'].fillna(X2_train['capital_loss'].mode()[0], inplace=True)
    salary_train2['hours_per_week'].fillna(X2_train['hours_per_week'].mode()[0], inplace=True)


In [ ]:
X2_train.isnull().sum()

age               0
workclass         0
education         0
education_num     0
marital_status    0
occupation        0
relationship      0
race              0
sex               0
capital_gain      0
capital_loss      0
hours_per_week    0
native_country    0
dtype: int64

**Encode categorical variable**

In [ ]:
# print categorical variables

categorical

['workclass',
 'education',
 'marital_status',
 'occupation',
 'relationship',
 'race',
 'sex',
 'native_country']

In [ ]:
X1_train[categorical].head()

workclass      education       marital_status       occupation  \
26840   Private   Some-college   Married-civ-spouse   Prof-specialty   
23346   Private      Bachelors        Never-married   Prof-specialty   
27598   Private      Bachelors              Widowed    Other-service   
15646   Private      Bachelors        Never-married            Sales   
7455    Private   Some-college   Married-civ-spouse    Other-service   

         relationship                 race      sex  native_country  
26840         Husband                White     Male   United-States  
23346       Own-child   Asian-Pac-Islander   Female   United-States  
27598   Not-in-family                White   Female   United-States  
15646       Own-child                White     Male   United-States  
7455          Husband                White     Male   United-States

In [ ]:
X2_train[categorical].head()

workclass      education       marital_status          occupation  \
1514     Private   Some-college   Married-civ-spouse        Craft-repair   
2932   Local-gov        Masters   Married-civ-spouse     Exec-managerial   
3284     Private   Some-college             Divorced        Craft-repair   
582      Private        HS-grad   Married-civ-spouse   Machine-op-inspct   
137      Private   Some-college        Never-married       Other-service   

        relationship    race    sex  native_country  
1514         Husband   White   Male   United-States  
2932         Husband   White   Male   United-States  
3284       Own-child   White   Male   United-States  
582          Husband   White   Male   United-States  
137    Not-in-family   Black   Male   United-States

In [ ]:
!pip install category_encoders

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 72 kB 768 kB/s 


In [ ]:
import category_encoders as ce

In [ ]:
# encode remaining variables with one-hot encoding

encoder=ce.OneHotEncoder(cols=['workclass', 'education', 'marital_status', 'occupation', 'relationship', 'race', 'sex', 'native_country'])

X1_train= encoder.fit_transform(X1_train)

X1_test= encoder.transform(X1_test)

In [ ]:
X1_train.head()

age  workclass_1  workclass_2  workclass_3  workclass_4  workclass_5  \
26840   55            1            0            0            0            0   
23346   26            1            0            0            0            0   
27598   31            1            0            0            0            0   
15646   27            1            0            0            0            0   
7455    40            1            0            0            0            0   

       workclass_6  workclass_7  education_1  education_2  ...  \
26840            0            0            1            0  ...   
23346            0            0            0            1  ...   
27598            0            0            0            1  ...   
15646            0            0            0            1  ...   
7455             0            0            1            0  ...   

       native_country_31  native_country_32  native_country_33  \
26840                  0                  0                  0   
23346                  0                  0                  0   
27598                  0                  0                  0   
15646                  0                  0                  0   
7455                   0                  0                  0   

       native_country_34  native_country_35  native_country_36  \
26840                  0                  0                  0   
23346                  0                  0                  0   
27598                  0                  0                  0   
15646                  0                  0                  0   
7455                   0                  0                  0   

       native_country_37  native_country_38  native_country_39  \
26840                  0                  0                  0   
23346                  0                  0                  0   
27598                  0                  0                  0   
15646                  0                  0                  0   
7455                   0                  0                  0   

       native_country_40  
26840                  0  
23346                  0  
27598                  0  
15646                  0  
7455                   0  

[5 rows x 102 columns]

In [ ]:
X1_train.shape

(24128, 102)

In [ ]:
X1_test.head()

age  workclass_1  workclass_2  workclass_3  workclass_4  workclass_5  \
25338   21            1            0            0            0            0   
18840   21            1            0            0            0            0   
8391    56            1            0            0            0            0   
18258   43            0            0            0            0            1   
16669   53            0            0            0            0            0   

       workclass_6  workclass_7  education_1  education_2  ...  \
25338            0            0            0            0  ...   
18840            0            0            0            0  ...   
8391             0            0            0            1  ...   
18258            0            0            1            0  ...   
16669            1            0            0            1  ...   

       native_country_31  native_country_32  native_country_33  \
25338                  0                  0                  0   
18840                  0                  0                  0   
8391                   0                  0                  0   
18258                  0                  0                  0   
16669                  0                  0                  0   

       native_country_34  native_country_35  native_country_36  \
25338                  0                  0                  0   
18840                  0                  0                  0   
8391                   0                  0                  0   
18258                  0                  0                  0   
16669                  0                  0                  0   

       native_country_37  native_country_38  native_country_39  \
25338                  0                  0                  0   
18840                  0                  0                  0   
8391                   0                  0                  0   
18258                  0                  0                  0   
16669                  0                  0                  0   

       native_country_40  
25338                  0  
18840                  0  
8391                   0  
18258                  0  
16669                  0  

[5 rows x 102 columns]

In [ ]:
X1_test.shape

(6033, 102)

In [ ]:
# encode remaining variables with one-hot encoding

encoder=ce.OneHotEncoder(cols=['workclass', 'education', 'marital_status', 'occupation', 'relationship', 'race', 'sex', 'native_country'])

X2_train= encoder.fit_transform(X2_train)

X2_test= encoder.transform(X2_test)

In [ ]:
X2_train.head()

age  workclass_1  workclass_2  workclass_3  workclass_4  workclass_5  \
1514   43            1            0            0            0            0   
2932   57            0            1            0            0            0   
3284   42            1            0            0            0            0   
582    41            1            0            0            0            0   
137    28            1            0            0            0            0   

      workclass_6  workclass_7  education_1  education_2  ...  \
1514            0            0            1            0  ...   
2932            0            0            0            1  ...   
3284            0            0            1            0  ...   
582             0            0            0            0  ...   
137             0            0            1            0  ...   

      native_country_31  native_country_32  native_country_33  \
1514                  0                  0                  0   
2932                  0                  0                  0   
3284                  0                  0                  0   
582                   0                  0                  0   
137                   0                  0                  0   

      native_country_34  native_country_35  native_country_36  \
1514                  0                  0                  0   
2932                  0                  0                  0   
3284                  0                  0                  0   
582                   0                  0                  0   
137                   0                  0                  0   

      native_country_37  native_country_38  native_country_39  \
1514                  0                  0                  0   
2932                  0                  0                  0   
3284                  0                  0                  0   
582                   0                  0                  0   
137                   0                  0                  0   

      native_country_40  
1514                  0  
2932                  0  
3284                  0  
582                   0  
137                   0  

[5 rows x 103 columns]

In [ ]:
X2_train.shape

(5238, 103)

In [ ]:
X2_test.head()

age  workclass_1  workclass_2  workclass_3  workclass_4  workclass_5  \
916    23            1            0            0            0            0   
5281   21            1            0            0            0            0   
1551   29            1            0            0            0            0   
4868   24            0            1            0            0            0   
1643   34            0            1            0            0            0   

      workclass_6  workclass_7  education_1  education_2  ...  \
916             0            0            1            0  ...   
5281            0            0            1            0  ...   
1551            0            0            0            0  ...   
4868            0            0            0            0  ...   
1643            0            0            0            0  ...   

      native_country_31  native_country_32  native_country_33  \
916                   0                  0                  0   
5281                  0                  0                  0   
1551                  0                  0                  0   
4868                  0                  0                  0   
1643                  0                  0                  0   

      native_country_34  native_country_35  native_country_36  \
916                   0                  0                  0   
5281                  0                  0                  0   
1551                  0                  0                  0   
4868                  0                  0                  0   
1643                  0                  0                  0   

      native_country_37  native_country_38  native_country_39  \
916                   0                  0                  0   
5281                  0                  0                  0   
1551                  0                  0                  0   
4868                  0                  0                  0   
1643                  0                  0                  0   

      native_country_40  
916                   0  
5281                  0  
1551                  0  
4868                  0  
1643                  0  

[5 rows x 103 columns]

In [ ]:
X2_test.shape

(1310, 103)

**Feature Scaling**

In [ ]:
cols=X1_train.columns

In [ ]:
from sklearn.preprocessing import RobustScaler

scaler = RobustScaler()

X1_train = scaler.fit_transform(X1_train)

X1_test = scaler.transform(X1_test)

In [ ]:
X1_train=pd.DataFrame(X1_train,columns=[cols])

In [ ]:
X1_test=pd.DataFrame(X1_test,columns=[cols])

In [ ]:
X1_train.head()

age workclass_1 workclass_2 workclass_3 workclass_4 workclass_5  \
0  0.947368         0.0         0.0         0.0         0.0         0.0   
1 -0.578947         0.0         0.0         0.0         0.0         0.0   
2 -0.315789         0.0         0.0         0.0         0.0         0.0   
3 -0.526316         0.0         0.0         0.0         0.0         0.0   
4  0.157895         0.0         0.0         0.0         0.0         0.0   

  workclass_6 workclass_7 education_1 education_2  ... native_country_31  \
0         0.0         0.0         1.0         0.0  ...               0.0   
1         0.0         0.0         0.0         1.0  ...               0.0   
2         0.0         0.0         0.0         1.0  ...               0.0   
3         0.0         0.0         0.0         1.0  ...               0.0   
4         0.0         0.0         1.0         0.0  ...               0.0   

  native_country_32 native_country_33 native_country_34 native_country_35  \
0               0.0               0.0               0.0               0.0   
1               0.0               0.0               0.0               0.0   
2               0.0               0.0               0.0               0.0   
3               0.0               0.0               0.0               0.0   
4               0.0               0.0               0.0               0.0   

  native_country_36 native_country_37 native_country_38 native_country_39  \
0               0.0               0.0               0.0               0.0   
1               0.0               0.0               0.0               0.0   
2               0.0               0.0               0.0               0.0   
3               0.0               0.0               0.0               0.0   
4               0.0               0.0               0.0               0.0   

  native_country_40  
0               0.0  
1               0.0  
2               0.0  
3               0.0  
4               0.0  

[5 rows x 102 columns]

In [ ]:
cols=X2_train.columns

In [ ]:
from sklearn.preprocessing import RobustScaler

scaler = RobustScaler()

X2_train = scaler.fit_transform(X2_train)

X2_test = scaler.transform(X2_test)

In [ ]:
X2_train=pd.DataFrame(X2_train,columns=[cols])

In [ ]:
X2_test=pd.DataFrame(X2_test,columns=[cols])

In [ ]:
X2_train.head()

age workclass_1 workclass_2 workclass_3 workclass_4 workclass_5  \
0  0.315789         0.0         0.0         0.0         0.0         0.0   
1  1.052632        -1.0         1.0         0.0         0.0         0.0   
2  0.263158         0.0         0.0         0.0         0.0         0.0   
3  0.210526         0.0         0.0         0.0         0.0         0.0   
4 -0.473684         0.0         0.0         0.0         0.0         0.0   

  workclass_6 workclass_7 education_1 education_2  ... native_country_31  \
0         0.0         0.0         1.0         0.0  ...               0.0   
1         0.0         0.0         0.0         1.0  ...               0.0   
2         0.0         0.0         1.0         0.0  ...               0.0   
3         0.0         0.0         0.0         0.0  ...               0.0   
4         0.0         0.0         1.0         0.0  ...               0.0   

  native_country_32 native_country_33 native_country_34 native_country_35  \
0               0.0               0.0               0.0               0.0   
1               0.0               0.0               0.0               0.0   
2               0.0               0.0               0.0               0.0   
3               0.0               0.0               0.0               0.0   
4               0.0               0.0               0.0               0.0   

  native_country_36 native_country_37 native_country_38 native_country_39  \
0               0.0               0.0               0.0               0.0   
1               0.0               0.0               0.0               0.0   
2               0.0               0.0               0.0               0.0   
3               0.0               0.0               0.0               0.0   
4               0.0               0.0               0.0               0.0   

  native_country_40  
0               0.0  
1               0.0  
2               0.0  
3               0.0  
4               0.0  

[5 rows x 103 columns]

**Model Training**

In [ ]:
# train a Gaussian Naive Bayes classifier on the training set
from sklearn.naive_bayes import GaussianNB

In [ ]:
# instantiate the model
gnb = GaussianNB()

In [ ]:
# fit the model
gnb.fit(X1_train, y1_train)

GaussianNB()

**Predict the result**

In [ ]:
y1_pred = gnb.predict(X1_test)

y1_pred

array([' <=50K', ' <=50K', ' <=50K', ..., ' <=50K', ' <=50K', ' <=50K'],
      dtype='<U6')

**Accuracy score**

In [ ]:
from sklearn.metrics import accuracy_score

print('Model accuracy score: {0:0.4f}'. format(accuracy_score(y1_test, y1_pred)))

Model accuracy score: 0.7989


Compare the train-set and test-set accuracy

In [ ]:
y1_pred_train = gnb.predict(X1_train)

y1_pred_train

array([' >50K', ' <=50K', ' <=50K', ..., ' <=50K', ' >50K', ' <=50K'],
      dtype='<U6')

In [ ]:
print('Training-set accuracy score: {0:0.4f}'. format(accuracy_score(y1_train, y1_pred_train)))

Training-set accuracy score: 0.8021


Check for overfitting and underfitting

In [ ]:
# print the scores on training and test set

print('Training set score: {:.4f}'.format(gnb.score(X1_train, y1_train)))

print('Test set score: {:.4f}'.format(gnb.score(X1_test, y1_test)))

Training set score: 0.8021
Test set score: 0.7989


The training-set accuracy score is 0.8021 while the test-set accuracy to be 0.7989. These two values are quite comparable. So, there is no sign of overfitting.

In [ ]:
# check class distribution in test set

y1_test.value_counts()

 <=50K    4559
 >50K     1474
Name: salary, dtype: int64

In [ ]:
# check null accuracy score

null_accuracy = (4559/(4559+1474))

print('Null accuracy score: {0:0.4f}'. format(null_accuracy))

Null accuracy score: 0.7557


We can see that our model accuracy score is 0.7989 but null accuracy score is 0.7557. So, we can conclude that our Gaussian Naive Bayes Classification model is doing a very good job in predicting the class labels.